app/common/request.server.ts

```diff
@@ -39,3 +73,41 @@ export const createApiRequest = (baseUrl: string) => {
     remove,
   };
 };
+
+export async function handleApiResponse<Data>(resp: Response) {
+  if (!resp.ok) {
+    let responseBody: any = await resp.text();
+    try {
+      responseBody = JSON.parse(responseBody);
+    } catch (e) {
+      // Ignore, we'll just return the text body
+    }
+
+    const result: ApiResult<null> = {
+      success: false,
+      status: resp.status,
+      errors: responseBody?.errors || [{ message: "Unable to fetch: " + resp.url }],
+    };
+    throw json(result, resp.status);
+  }
+
+  const responseBody: ApiResponseBody<Data> = await resp.json().catch((err) => {
+    console.error("Unable to parse response as JSON", err);
+    throw json(
```

Comment:
This is interesting... 

I tried throwing the ApiResult object, but remix wasn't picking it up in the Error boundary. It was just picking up an empty error. Leading me to believe you HAVE to throw actual errors to leverage the error boundary.  

However, remix also has a `CatchBoundary` which will catch a `Response` you throw anywhere in the call stack.